File: F# Workspace/04_Tuples/04_Tuples/time_components.fs
(*
 * time_components.fs
 * By Antonio F. Huertas
 * This program prints the time components given the elapsed seconds.
 *)

let get_components elapsed_secs =
    let secs_per_hour = 3600
    let hrs = elapsed_secs / secs_per_hour
    let rem_secs = elapsed_secs % secs_per_hour

    let secs_per_min = 60
    let mins = rem_secs / secs_per_min
    let secs = rem_secs % secs_per_min

    (hrs, mins, secs)

// start : int -> unit
// Starts the execution of the program.
let start elapsed_secs =
    printfn "The number of elapsed seconds is %d" elapsed_secs

    if elapsed_secs < 0 then
        printfn "Error! The number of elapsed seconds is negative."
    else
        let (hrs, mins, secs) = get_components elapsed_secs
        printfn "\nAmount of time is %d hours, %d minutes and %d seconds" 
                hrs mins secs

File: C_Projects/02_Pointers/time_components.c
/*
* time_components.c
* this program determines the components (hours, minutes, seconds)
* for the given number of elapsed seconds*/

#include <stdio.h>  //for printf, scanf

//Prototype
void get_components(int elapsed_secs, int *hrs, int *mins, int *secs);

// Starts the execution of this program
int main(void) {
    int elapsed_seconds;

    printf("Enter the number of elapsed seconds: ");
    scanf("%d", &elapsed_seconds);

    if(elapsed_seconds < 0)
        printf("Error! Negative elapsed seconds. \n");
        else {
            int hours, minutes, seconds;
            get_components(elapsed_seconds, &hours, &minutes, &seconds);
            printf("The components are %d hours, %d minutes and %d seconds.\n", hours, minutes, seconds);
        }
    return 0;
}

void get_components(int elapsed_secs, int *hrs, int *mins, int *secs) {
    const int SECS_PER_MIN = 60, SECS_PER_HOUR = 3600;

    *hrs = elapsed_secs / SECS_PER_HOUR;
    int rem_secs = elapsed_secs % SECS_PER_HOUR;
    *mins = rem_secs / SECS_PER_MIN;
    *secs = rem_secs % SECS_PER_MIN;
}
